### ISSUE TITLE
ML primitives roadmap: FFT, pooling, embedding, attention helpers

### ISSUE THREAD

**Issue by noahgift** [paiml/trueno#61]
## Summary

Add missing ML primitives to trueno for comprehensive neural network support in the Sovereign AI Stack.

## Current Capabilities ✅

trueno already has excellent coverage:

### Vector Operations
- Arithmetic: add, sub, mul, div, dot, sum, scale, fma
- Statistics: mean, variance, stddev, covariance, correlation, zscore
- Norms: L1, L2, Linf, normalize
- Activations: relu, leaky_relu, elu, gelu, selu, sigmoid, tanh, softmax, log_softmax, swish, mish, hardswish
- Math: exp, ln, log2, log10, sqrt, pow, sin, cos, tan, sinh, cosh, tanh, floor, ceil, round
- Utilities: argmax, argmin, clip, clamp, abs, neg, minimum, maximum

### Matrix Operations
- matmul, transpose, matvec, vecmat
- conv2d (with GPU shader)
- identity, zeros

### Tier 1: Critical for WASM ML Demos

| Operation | Use Case | Effort |
|-----------|----------|--------|
| **embedding_lookup** | NLP models (word → vector) | 2-3 hrs |
| **max_pool2d** | CNNs (downsampling) | 3-4 hrs |
| **avg_pool2d** | CNNs (downsampling) | 2-3 hrs |
| **conv1d** | Sequence models, audio | 4-5 hrs |
| **layer_norm** | Transformers | 2-3 hrs |
| **batch_norm** | CNNs (inference mode) | 2-3 hrs |

### Tier 2: Needed for Advanced Models

| Operation | Use Case | Effort |
|-----------|----------|--------|
| **bmm** (batch matmul) | Attention mechanisms | 4-6 hrs |
| **topk** | Beam search, sampling | 2-3 hrs |
| **argsort** | Ranking, top-k | 2-3 hrs |
| **gather / scatter** | Embedding, indexing | 3-4 hrs |
| **pad** | Conv padding modes | 2-3 hrs |
| **upsample** | Image generation | 3-4 hrs |

### Tier 3: Signal Processing (for Whisper/Audio)

| Operation | Use Case | Effort |
|-----------|----------|--------|
| **fft** | Mel spectrogram, audio | 8-12 hrs |
| **ifft** | Audio reconstruction | 4-6 hrs |
| **stft** | Short-time FFT | 4-6 hrs |
| **mel_filterbank** | Audio features | 3-4 hrs |

### embedding_lookup
```rust
/// Lookup embeddings by indices
/// table: [vocab_size, embed_dim], indices: [seq_len] → output: [seq_len, embed_dim]
pub fn embedding_lookup(table: &Matrix<f32>, indices: &[usize]) -> Matrix<f32>
```

### max_pool2d / avg_pool2d
```rust
/// 2D max pooling with kernel size and stride
pub fn max_pool2d(input: &Matrix<f32>, kernel: (usize, usize), stride: (usize, usize)) -> Matrix<f32>
```

### layer_norm
```rust
/// Layer normalization: (x - mean) / sqrt(var + eps) * gamma + beta
pub fn layer_norm(input: &Vector<f32>, gamma: &Vector<f32>, beta: &Vector<f32>, eps: f32) -> Vector<f32>
```

### FFT (Cooley-Tukey)
```rust
/// Radix-2 FFT for power-of-2 lengths
/// Returns (real, imag) components
pub fn fft(input: &Vector<f32>) -> (Vector<f32>, Vector<f32>)
```

## Priority for Presentar WASM Demos

For immediate .apr demos in presentar:
1. **embedding_lookup** - enables NLP demos
2. **layer_norm** - enables transformer-style models
3. **conv1d** - enables sequence models
4. **max_pool2d** - enables CNN demos (MNIST)

## References

- [FFT Cooley-Tukey](https://en.wikipedia.org/wiki/Cooley%E2%80%93Tukey_FFT_algorithm)
- [Layer Normalization](https://arxiv.org/abs/1607.06450) - Ba et al. 2016
- [Batch Normalization](https://arxiv.org/abs/1502.03167) - Ioffe & Szegedy 2015

## Labels

`enhancement`, `P2`, `ml-primitives`

**Comment by noahgift** [comment]
✅ Implemented `embedding_lookup` (Tier 1 item #1)

- `Matrix::embedding_lookup(&self, indices: &[usize])` - basic lookup
- `Matrix::embedding_lookup_sparse(&self, indices: &[usize])` - returns unique indices for sparse gradient updates
- 7 comprehensive tests + doctest
- See commit 88672a6

Remaining Tier 1 items:
- [ ] max_pool2d
- [ ] avg_pool2d
- [ ] conv1d
- [ ] layer_norm
- [ ] batch_norm

**Comment by noahgift** [comment]
✅ Implemented `layer_norm` (Tier 1 item #2)

- `Vector::layer_norm(&self, gamma, beta, eps)` - full layer norm with learnable params
- `Vector::layer_norm_simple(&self, eps)` - simplified version without params
- 9 unit tests + 2 doctests
- See commit 9632bad

Remaining Tier 1 items:
- [x] embedding_lookup ✅
- [x] layer_norm ✅
- [ ] max_pool2d
- [ ] avg_pool2d
- [ ] conv1d
- [ ] batch_norm

**Comment by noahgift** [comment]
## Status Update (2026-01-03)

### Already Implemented ✅

| Operation | Location | Notes |
|-----------|----------|-------|
| `embedding_lookup` | `matrix.rs:2189` | Full SIMD implementation |
| `embedding_lookup_sparse` | `matrix.rs:2240` | Sparse variant |
| `layer_norm` | `vector.rs:1316` | With learnable gamma/beta |
| `layer_norm_simple` | `vector.rs:1386` | Without learnable params |
| `batched_matmul` | `matrix.rs:377` | For 3D tensors |
| `batched_matmul_4d` | `matrix.rs:458` | For attention (batch×heads) |

### Remaining Tier 1 Items

- [ ] `max_pool2d` / `avg_pool2d` - CNNs
- [ ] `conv1d` - Sequence models
- [ ] `batch_norm` (inference) - CNNs

### Remaining Tier 2/3 Items

- [ ] `topk`, `argsort` - Beam search
- [ ] `gather`, `scatter` - Indexing
- [ ] `pad`, `upsample` - Conv helpers
- [ ] FFT family (complex computation, lower priority)

Most Tier 1 primitives are now implemented. Keeping this open to track remaining items.

**Comment by noahgift** [comment]
## Commit 62ed207 - Additional ML Primitives

### Newly Implemented ✅

| Operation | Location | Notes |
|-----------|----------|-------|
| `max_pool2d` | `matrix.rs:2277` | 2D max pooling with kernel/stride |
| `avg_pool2d` | `matrix.rs:2340` | 2D average pooling |
| `topk` | `matrix.rs:2394` | Top-K selection for beam search |
| `gather` | `matrix.rs:2430` | Index selection along axis |
| `pad` | `matrix.rs:2486` | Constant padding |

### Test Coverage

9 new tests added:
- `test_max_pool2d`, `test_max_pool2d_stride_1`
- `test_avg_pool2d`
- `test_topk`, `test_topk_empty`
- `test_gather_rows`, `test_gather_cols`
- `test_pad`, `test_pad_asymmetric`

### Remaining Items

- [ ] `conv1d` - Sequence models
- [ ] `batch_norm` (inference) - CNNs
- [ ] FFT family - Lower priority, complex computation

**Comment by noahgift** [comment]
## Progress Update (2026-01-03)

### Tier 2 Primitives Added ✅

The following ML primitives have been implemented in `src/matrix.rs`:

| Operation | Status | Tests |
|-----------|--------|-------|
| `max_pool2d` | ✅ Implemented | 3 tests |
| `avg_pool2d` | ✅ Implemented | 2 tests |
| `topk` | ✅ Implemented | 2 tests |
| `gather` | ✅ Implemented | 2 tests |
| `pad` | ✅ Implemented | 2 tests |

### API Examples

```rust
// 2D max pooling
let pooled = matrix.max_pool2d((2, 2), (2, 2))?;

// 2D average pooling
let pooled = matrix.avg_pool2d((2, 2), (2, 2))?;

// Top-k values and indices
let (values, indices) = matrix.topk(5)?;

// Gather rows by indices
let gathered = matrix.gather(&[0, 2, 4], 0)?;

// Pad with constant value
let padded = matrix.pad(((1, 1), (1, 1)), 0.0)?;
```

### Remaining Work (Tier 1, 3)

- [ ] `embedding_lookup` - NLP models
- [ ] `layer_norm` - Transformers
- [ ] `batch_norm` - CNNs (inference mode)
- [ ] `conv1d` - Sequence models
- [ ] `fft` / `ifft` / `stft` - Signal processing
- [ ] `mel_filterbank` - Audio features

This issue remains open to track the roadmap.
